Support Up To 3 Bluetooth Devices

hftsai256 opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

* Correct DIP switch logic on v1.4 board
* Support up to 3 bluetooth devices as supported by original fw
* Optimize UART comm buffer with BLE module
* Remove redundant BLE commands
* Invoke modular functions from centralized event manager
* Capture minor hardware differences as board "traits" instead from
  upstreamed HAL library in `chibios-contrib`

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [x] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

(Not available)

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

hftsai256 committed:
correct pins

hftsai256 committed:
correct DIP switch logic

Codetector1374 commented on keyboards/kemove/boards/KEMOVE_SNOWFOX/v1_4/board.h:
```diff
@@ -68,9 +68,9 @@
-#define LINE_BLE_RSTN                                   PAL_LINE(IOPORT0, 20)
+#define LINE_UART_RX                                    PAL_LINE(IOPORT1, 26) // To pin 21 on BLE
+#define LINE_UART_TX                                    PAL_LINE(IOPORT1, 27) // To pin 20 on BLE
+#define LINE_BLE_RSTN                                   PAL_LINE(IOPORT0, 23) // Unsure, to pin 11 on BLE
```
This should be correct. It's been a long time since I last worked on this, but I am 99% sure there's only three connection going between the MCU and the BLE module. This should be the reset pin.

Codetector1374 commented on keyboards/kemove/snowfox/keymaps/default/keymap.c:
```diff
@@ -126,7 +126,7 @@ const uint16_t keymaps_size = sizeof(keymaps);
   switch (index) {
     case 0:
-      if (active) {
+      if (!active) {
```
Wouldn't this also affect other revision's function?

hftsai256 replied:
I think you are correct. Then the second question would be how to manage this difference within the abstraction layer. I would propose leveraging the inverse function built into GPIOs by complementing the HAL header in `chibios-contrib` to add the following macros:
```
#define   MODE_HYS_POS            5U
#define   MODE_HYS_DISABLE        (0x0U << MODE_HYS_POS)
#define   MODE_HYS_ENABLE         (0x1U << MODE_HYS_POS)

#define   MODE_INV_POS            6U
#define   MODE_INV_DISABLE        (0x0U << MODE_INV_POS)
#define   MODE_INV_ENABLE         (0x1U << MODE_INV_POS)
```

So that the inverse logic is dealt within the board initialization code:
```
    palSetLineMode(LINE_WINMAC, MODE_DIR_IN | MODE_MODE_PULL_UP | MODE_INV_ENABLE | MODE_AD_DIGITAL);
```

Does this sound feasible?

Codetector1374 replied:
I think the solution you proposed is quite elegant

You can make the change to ChibiOS if you'd like and test it locally make sure it works. I've never tested the INVERT function before. But I'd assume it works. 

If it does not work, I can try and read QMK to see if there is something that might help us in this case.

hftsai256 replied:
I had a local build with the proposal above, and fortunately it works pretty well.

I believe QMK should have such an abstracted interface for such a thing, but if our journey can stop here then I wouldn't bother too much about it.

Codetector1374 replied:
I think it is fine to leave it like this if it works.

hftsai256 replied:
One thing just came across my mind: Can we confirm that the current logic works across v1.2 v1.5 and v3 except for v1.4? That sounds a bit weird to me TBH

Codetector1374 replied:
Yeah this is interesting, can't say anything for 1.5, but I have 1.2 and v3, they both work correctly

hftsai256 replied:
Well if that's the case then I'll push the change. Am I supposed to create another PR on `chibios-contrib` repo?

Codetector1374 replied:
I think that needs to happen first.

hftsai256 replied:
It's opened [here](https://github.com/OpenKemove/ChibiOS-Contrib/pull/1), but for some reason I couldn't add you as a reviewer.

hftsai256 committed:
absorb DIP logic to board specific section

hftsai256 committed:
Expend bluetooth features to cover what original stack firmware covers.

* Correct DIP switch logic (v1.4)
* Support up to 3 bluetooth devices as supported by original fw
* Optimize UART comm buffer with BLE module
* Remove redundant BLE commands
* Invoke modular functions from centralized event manager
* Capture minor hardware differences as board "traits" instead from
  upstreamed HAL library in `chibios-contrib`

hftsai256 commented:
I moved the required HAL update into a board-specific constant variable `traits`, and also explicitly assigned three keys to switch between bluetooth devices.

One special note is that the BLE module will respond `OK` for almost everything but `AT+HID`, and the stack firmware doesn't seem to respect this `OK` acknowledgement as it will send the next command without the current one being cleared. I was considering implementing a ring buffer task queue to wait until a configuration command is cleared by an `OK`, but on the other hand, what are we going to deal with it if a command timed out? Possible options are (based on features we can do at this time):
1. Re-schedule this task only.
2. Re-schedule all tasks following the failed task.
3. Do nothing. This doesn't matter as long as it is getting `+EVENT=` response, which is the one we should really care about.

Ideally we want a user-feedback through LED lighting. Therefore I would switch gear toward LED module and see if there is a proper way to send events from main/BLE threads with a flag to request a specific key to blink.

hftsai256 committed:
Strip out BLE semaphore

Remove BLE 'OK' state semaphore as it wouldn't being accessed by other
threads. On the other hand, because sdGetTimeout is a blocking read,
ble_command_withOK() will always get timed out.

The proper way of doing this is to create a task for each uart tx
transmission, and clear/join the task when receiving an OK on
first-come-first-serve basis.

v1.4 code is merged into v1.5 as they share the same pin configuration.

Codetector1374 commented:
I think the change looks good to me. The traits structure make sense in this context. Regarding the `OK`, I originally considered it as well, but the problem is timing could be annoying, I guess putting the command in a queue and some thread consuming it is probably the best solution there. On error we could either retry or just give up after let's say 10ms. I think for a keyboard people do expect it to respond within a reasonable amount of time, if after that the expectation here is that "that press didn't work, I should press again"
